### DISCUSSION TITLE
Can R package vignettes be written in Quarto?

### DISCUSSION THREAD

**Discussion by dcnorris** [quarto-dev/quarto-cli#2307] [Q&A]
I see that even the **quarto** package has [an Rmd vignette](https://github.com/quarto-dev/quarto-r/blob/main/vignettes/publishing.Rmd), so I assume the answer is "not yet".

Are there plans to support Quarto vignettes? Might a `quarto` VignetteEngine be provided to replace `knitr::rmarkdown`? Or are are there deeper issues, e.g. aspects of Quarto's design that are incompatible with R package vignettes?

**Comment by mcanouil** [comment]
I don't think it's going to happen in a near future because CRAN would need to add Quarto binary to their setup.
So, this is entirely up to the R Core Team.

**Reply by dcnorris** [reply]
It's better to let CRAN catch up with advances than to give it veto power over them. CRAN has operational constraints and design aims that (perhaps quite appropriately?) are incompatible with rapid progress. Adding compiled Rust code to my R package [precautionary](https://dcnorris.github.io/precautionary/), necessitated that I abandon CRAN — for the time being, at least.

The best way to promote Quarto vignettes would be to 'build it and let them come'. After just a few days with Quarto, I already feel about Rmd the same way I felt about Sweave when I made that switch. I have little appetite to write another Rmd vignette!

**Reply by mcanouil** [reply]
We should give it a bit of time.
Quarto is only ~ 1.5 month old (birth: 28th of July 2022) 😅

Quarto makes more sense to me as Rmarkdown never has (I am not even talking about Sweave 😆).

**Comment by dcnorris** [comment]
Actually, this isn't nearly as hard as I imagined. Thanks to the example provided by @bdcaf's https://github.com/bdcaf/vignetteEngineMake, I see it's fairly straightforward to add Quarto vignettes to my package:

1. In the `DESCRIPTION` file, add your own package to the `VignetteEngines` line (let's say mine is called TOPSECRET):
```
VignetteBuilder: knitr, rmarkdown, TOPSECRET
```
2. Add an `.onLoad` function to your package (e.g., in a `zzz.R` file), to register a `quarto` vignetteEngine:
```r
.onLoad <- function(libname, pkgname) {
  tools::vignetteEngine(name = "quarto",
                        package = pkgname,
                        pattern = "[.]qmd$",
                        weave = function(file, ..., encoding = "UTF-8") {
                          message("Hello from custom Quarto vignette engine!")
                          ## TODO: What is Quarto's expectation about encoding?
                          ## NB: output_format = "all" below might make a better default
                          quarto::quarto_render(file, ..., output_format = "pdf")
                        },
                        tangle = tools::vignetteEngine("knitr::rmarkdown")$tangle,
                        aspell = tools::vignetteEngine("knitr::rmarkdown")$aspell
                        )
}
```
3. In the YAML header for the Quarto vignette, specify the qualified vignetteEngine:
```
vignette: >
  %\VignetteIndexEntry{Whatever}
  %\VignetteEngine{TOPSECRET::quarto}
  %\VignetteEncoding{UTF-8}
```

This makes a perfectly workable solution for me. But I believe a cleaner way to do this would be to define several vignetteEngines in the `quarto` package — say `quarto::pdf`, `quarto::html` and even `quarto::all`. Then instead of the awkward self-reference in the VignetteBuilder line, you would simply write `quarto`.

**Reply by mcanouil** [reply]
The issue is not about registering the vignette engine.
This won't work on CRAN unless Quarto (not the R package) is installed on their server, especially for Windows version of the R packages.

The `rmarkdown` package rely on R which is obviously available for CRAN checks.
`quarto` is a wrapper to the Quarto CLI binary, without the library, the R package `quarto` is useless.

**Reply by dcnorris** [reply]
CRAN is merely one of several ways to host an R package, and the R package mechanism is in any case more broadly useful (for abstraction and whatnot) than merely a means to post code online. Anyone building a package for personal use, for a commercial client (my own use case here), or indeed for sharing *anywhere* other than CRAN, could well find my suggested changes to `quarto-r` useful.

**Reply by mcanouil** [reply]
I did not say, it's not a useful trick.
I'd rather have my R packages on r-universe.dev in some aspects.
The main "issue" remains: Quarto is really really young as of right now.

If you want, you can propose your change as a PR in <https://github.com/quarto-dev/quarto-r>.

**Reply by dcnorris** [reply]
Initiated https://github.com/quarto-dev/quarto-r/pull/57

**Reply by HenrikBengtsson** [reply]
The PR https://github.com/quarto-dev/quarto-r/pull/57 by @dcnorris LGTM. I've just commented in that PR with a real-world example, confirming that it works correctly with `R CMD check --as-cran ...`

I think it's a good idea to add this to the **quarto** package, even if CRAN still doesn't have `quarto` binaries installs.  As soon as CRAN decides to install that, it'll work out of the box.  PS. The R Core has no say in what CRAN should do.

**Reply by cderv** [reply]
Thanks for sharing @dcnorris and confirming @HenrikBengtsson ! I plan to work on it this week so this is rightly timed. 
Idea is to have a minimal HTML version as vignette results. We don't want all bootstrap in a vignette HTML I believe. 
Work will happen in **quarto** R package indeed - we can discuss there too

**Reply by eitsupi** [reply]
Related to this thread

https://github.com/r-lib/pkgdown/issues/2210#issuecomment-1749733986

I saw a comment in the pkgdown repository that Quarto has been added to machines on CRAN.
Is this information publicly available somewhere? (Or will it be announced in the future?)

Rust was mentioned above (I have a package with Rust code on CRAN), there's really no way to know the Rust version on CRAN... (at least for me, the only thing I can rely on is the installation logs of packages using Rust).
I was just wondering if there is a way to find out about the existence of Quarto on CRAN in the same way.

**Reply by mcanouil** [reply]
By submitting a package with some tests failing and outputting the version etc.? ^^

**Reply by camilogarciabotero** [reply]
It seems that there is also an effort from python: https://github.com/machow/quartodoc

**Reply by mcanouil** [reply]
`quartodoc` is not exactly the same thing.
It's more a `roxygen2` + `pkgdown` tool but for Python.

**Reply by RaymondBalise** [reply]
Can someone please tell me the status on using Quarto for vignettes and CRAN?  I am teaching package development next semester, and I would love to focus on Quarto instead of R Markdown. Is @denorris's `.onLoad` function (wizardry) still the way to go? Has a package with Quarto based vignettes made it onto CRAN?

**Reply by dcnorris** [reply]
Latest progress described here: https://github.com/quarto-dev/quarto-r/pull/57#pullrequestreview-1712337041

**Reply by cderv** [reply]
A quarto vignette engine is almost ready. It will be merged soon in **quarto** R package.

**Reply by RaymondBalise** [reply]
Hello @cderv. Can you share your guestimate on when Quarto based vignettes will be ready to roll?  I am teaching package development this semester and I can, in theory, provide a solid test audience using Posit.Cloud if the timing is right.

**Reply by cderv** [reply]
This should be this week in v1.4 of Quarto R package that I am finalizing at : https://github.com/quarto-dev/quarto-r/

Tracked here. 
- https://github.com/quarto-dev/quarto-r/issues/136

**Comment by cderv** [comment]
For anyone who's interested and has not tested yet, the dev version of **quarto** R package now has some vignette engines registered. 

Please open new issues in https://github.com/quarto-dev/quarto-r for ideas, or any other feedback thanks!

**Reply by RaymondBalise** [reply]
Is there an example showing the YAML header to make an R package vignette with Quarto?  `usethis::use_vignette()` gives this header:

```
---
title: "demo"
output: rmarkdown::html_vignette
vignette: >
  %\VignetteIndexEntry{analysis}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
---
```

Is there a similar `usethis` function planned for quarto based vignettes?

**Reply by cderv** [reply]
For now **usethis** does not have a function for this. The example is given by Quarto own vignette and looking at demo vignette in the package (`vignette("hello", package = "quarto")`). 

This would be 
```` 
---
title: "demo"
vignette: >
  %\VignetteIndexEntry{demo}
  %\VignetteEngine{quarto::html}
  %\VignetteEncoding{UTF-8}
---
````

**pkgdown** does not handle Quarto vignette either, so they can't show on the website if they use Quarto feature - I'll create a new article not using quarto specifically for this I guess.

**Reply by tripartio** [reply]
It seems that pkgdown now supports Quarto vignettes: https://pkgdown.r-lib.org/articles/quarto.html

**Reply by LiNk-NY** [reply]
@cderv 
Hi Christophe!
Is it possible to render a package vignette using an `_extension` rather than `quarto::html`? 
I took a quick look at the `quarto-r` source code and I am not sure if that is the intention of `quarto::format` engine. 

I suppose the alternative is to register my own engine. I am not doing anything special just applying a `custom.scss` to my vignette but I want to be able to distribute this `custom.scss` (within an R package) to other members in my organization. 

Thank you in advance!

**Reply by cderv** [reply]
> Is it possible to render a package vignette using an _extension rather than quarto::html?
> I took a quick look at the quarto-r source code and I am not sure if that is the intention of quarto::format engine.


Yes - see this new vignette about the `quarto:format` engine: https://quarto-dev.github.io/quarto-r/articles/advanced-vignettes.html

However, you'll need a bit of configuration as `_extension` is a resource that you need to handle correctly for R CMD build to handle correctly when building a vignette. I must acknowledge I never tried with an extension I think 🤔 

> I am not doing anything special just applying a custom.scss to my vignette

So this is something like 
````
format: 
  html: 
    theme: custom.scss
```` 

right ? 

But you want this custom.scss to be in your package as an extension you would install? Or ship within the package ? Is this custom.scss used only for the vignette ? Or is this something you want to make available to others, and why you have a custom format ? 
````yaml
format: myorg-html
````

Please do try and open an issue in quarto-r repo if you have issues. I'll be happy to solve that with you. 
I would not advice to use extension for a CRAN vignette, but for a in-private organization package I can see the need; 

Thank you
